app/controllers/actions/MessageTypeIdentifierAction.scala

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.actions
+
+import com.google.inject.{ImplementedBy, Inject}
+import models.MessageType
+import models.requests.{MessageRecipientRequest, RoutableRequest}
+import play.api.Logger
+import play.api.mvc.{ActionBuilder, ActionRefiner, ActionTransformer, AnyContent, DefaultActionBuilder, Request, Result}
+import play.api.mvc.Results.BadRequest
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class MessageTypeIdentifierActionProvider @Inject()(
+     buildDefault: DefaultActionBuilder
+   )(implicit ec: ExecutionContext){
+  def apply(): ActionRefiner[MessageRecipientRequest, RoutableRequest] =
+    new MessageTypeIdentifierAction(ec)
+}
+
+class MessageTypeIdentifierAction(val ec: ExecutionContext)
+  extends ActionRefiner[MessageRecipientRequest, RoutableRequest] {
+
+  private lazy val logger = Logger(getClass)
+
+  override protected def refine[A](
+                                    request: MessageRecipientRequest[A]
+                                  ): Future[Either[Result, RoutableRequest[A]]] = {
+    println("start")
+      request.headers
+        .get("X-Message-Type") match {
+        case None => {
+          logger.debug("BadRequest: missing header key X-Message-Type")
+          Future.successful(Left(BadRequest("BadRequest: missing header key X-Message-Type")))
+        }
+        case Some(messageType) =>
+          MessageType.values.filter(x => x.code == messageType).headOption match {
```

Comment:
Find could be used here instead of filter to make the code more readable.<br>`MessageType.values.find(x => x.code == messageType)`.<br><br>The getMessageType method could also be amended and used here instead.